pages/agent/v3/signed-pipelines.md

```diff
@@ -0,0 +1,114 @@
+
+With regards to specific algorithm choice, basically any of the asymmetric signing algorithms we support are fine and will be secure. If you're not sure which one to use, `EdDSA` is proven to be secure, has a modern design, wasn't designed by a Nation State Actor, and produces nice short signatures.
+
+Signed pipelines supports the following signing algorithms:
+
+#### Symmetric
+- `HS256`
+- `HS384`
+- `HS512`
+
+#### Asymmetric
+- `PS256`
+- `PS384`
+- `PS512`
+- `ES256`
+- `ES384`
+- `ES512`
+- `EdDSA`
+
+### Making your agents sign steps they upload 
+
+Now that your keys are generated, we need to configure your agents to use them. On agents that will be uploading pipelines, add the following to your agent's config file:
+```ini
+job-signing-jwks-path=<path to signing keys>
+signing-key-id=<the key id you generated earlier>
+job-verification-jwks-path=<path to verification keys>
+```
+
+On instances that will verify jobs, add:
+```ini
+job-verification-jwks-path=<path to verification keys>
+```
+
+No
+
+### Signing the pipeline's stored configuration
+
+[ed: this section feels clunky to me, and i'm not sure the best way to explain it - we need to sign the initial steps of a pipeline as well as the steps that get uploaded by an agent, but it's difficult to concisely convey all the concepts involved]
+
+In Buildkite, pipelines have two YAML definitions - one that's stored Buildkite, which you can change on the settings page for a given pipeline, and other(s) that are uploaded by agents. In this guide, we'll refer to the former as the "stored" configuration, and the latter as "uploaded" configuration. The stored configuration generally contains the initial steps needed to run a pipeline, and almost always contains a step that uploads a pipeline to Buildkite.
+
+> Some older Buildkite pipelines don't use YAML for their stored pipeline configuration - these aren't supported by signed pipelines, and will have to be migrated to use the new format.
+
+You may have noticed that when we configured the signing instances, we gave them the capability to both sign and verify jobs. For maximum security, the stored pipeline will need to be
+
+The solution here is to add static signatures to the steps in your pipeline's stored configuration.
+
+[ed: explain how to do this - buildkite-agent pipeline upload --jwks-file-path /path/to/keys --signing-key-id blah --dry-run and copy and paste into buildkite's UI]
+
+### Hard mode: static signatures
+
+[ed: if you're really paranoid, you can sign a pipeline once, and hardcode the signatures into your pipeline yaml. this means that if anyone changes anything on the signed steps, your jobs will fail]
```

Comment:
What's the advantage to doing this? Is it so you don't have to trust agent updates that might introduce vulnerabilities/attacks via the signing process?

Reply:
Just wondering if it's worth even explaining here. You already explain signing stored configuration above - not that hard to connect the dots

Reply:
the advantage it offers over "dynamic" signing (ie signing the pipeline each time it's uploaded) is that it protects against the case of "someone got into my repo and turned all of my test scripts into bitcoin miners", which is definitely something i think at least some of our customers would appreciate. the downside is that basically every time you edit the pipeline.yml you'd have to re-sign it, which could get to be a pain

Reply:
given that we're going to be signing build invariants, this approach has a lot less utility - i'm gonna yeet it